pkg/kademlia/pslice/pslice.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package pslice
+
+import (
+	"sync"
+
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology"
+)
+
+// PSlice maintains a list of addresses, indexing them by their different proximity orders.
+// Currently, when peers are added or removed, their proximity order must be supplied, this is
+// in order to reduce duplicate PO calculation which is normally known and already needed in the
+// calling context.
+type PSlice struct {
+	peers []swarm.Address
+	bins  []uint
+
+	sync.Mutex
+}
+
+// New creates a new PSlice.
+func New(maxBins int) *PSlice {
+	return &PSlice{
+		peers: make([]swarm.Address, 0),
+		bins:  make([]uint, maxBins),
+	}
+}
+
+// iterates over all peers from deepest bin to shallowest.
+func (s *PSlice) EachBin(pf topology.EachPeerFunc) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if len(s.peers) == 0 {
+		return nil
+	}
+
+	var binEnd = uint(len(s.peers))
+
+	for i := len(s.bins) - 1; i >= 0; i-- {
+		peers := s.peers[s.bins[i]:binEnd]
+		for _, v := range peers {
+			stop, next, err := pf(v, uint8(i))
+			if err != nil {
+				return err
+			}
+			if stop {
+				return nil
+			}
+			if next {
+				break
+			}
+		}
+		binEnd = s.bins[i]
+	}
+
+	return nil
+}
+
+// EachBinRev iterates over all peers from shallowest to deepest.
+func (s *PSlice) EachBinRev(pf topology.EachPeerFunc) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if len(s.peers) == 0 {
+		return nil
+	}
+
+	var binEnd int
+	for i := 0; i <= len(s.bins)-1; i++ {
+		if i == len(s.bins)-1 {
+			binEnd = len(s.peers)
+		} else {
+			binEnd = int(s.bins[i+1])
+		}
+
+		peers := s.peers[s.bins[i]:binEnd]
+		for _, v := range peers {
+			stop, next, err := pf(v, uint8(i))
+			if err != nil {
+				return err
+			}
+			if stop {
+				return nil
+			}
+			if next {
+				break
+			}
+		}
+	}
+	return nil
+}
+
+// ShallowestEmpty returns the shallowest empty bin if one exists.
+// If such bin does not exists, returns true as bool value.
+// Must be called under lock.
+func (s *PSlice) ShallowestEmpty() (bin uint8, none bool) {
```

Comment:
this is an optimization. i'd be happy to do it when it is demonstrated that it gives an improvement. increasing complexity and amount of state we maintain on objects without justification is something i'd like to avoid. this is the same reason why this data structure is a simple slice and not some fancy `b+-!#^tree`.

Reply:
> i'd be happy to do it when it is demonstrated that it gives an improvement.

I don't think that needs a demonstration.

But putting it off till optimization stage is totally fine of course.